File: hot/leetcode/142.环形链表 II/142.环形链表 II.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if head is None:
            return head

        meet_point = self.find_meet_point(head)
        if meet_point is None:
            return None

        fast = meet_point
        slow = head
        while fast != slow:
            fast = fast.next
            slow = slow.next
        return slow

    def find_meet_point(self, head: ListNode) -> ListNode:
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return slow

        return None

File: hot/leetcode/142.环形链表 II/142.环形链表 II.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

/*
1.找到双指针相遇点
2.某一个指针指向head，然后同步前进，直到相等时为相遇点
*/
func detectCycle(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	meetPoint := findMeetPoint(head)
	if meetPoint == nil {
		return nil
	}
	fast := meetPoint
	slow := head
	for fast != slow {
		fast = fast.Next
		slow = slow.Next
	}
	return slow
}

func findMeetPoint(head *ListNode) *ListNode {
	fast := head
	slow := head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
		if fast == slow {
			return slow
		}
	}
	return nil
}
